src/passes/call-utils.h

```diff
@@ -0,0 +1,153 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef wasm_ir_function_h
+#define wasm_ir_function_h
+
+#include <variant>
+
+#include "ir/type-updating.h"
+#include "wasm.h"
+
+namespace wasm::CallUtils {
+
+// Define a variant to describe the information we know about an indirect call,
+// which is one of three things:
+//  * Unknown: Nothing is known this call.
+//  * Trap: This call target is invalid and will trap at runtime.
+//  * Known: This call goes to a known static call target, which is provided.
+struct Unknown : public std::monostate {};
+struct Trap : public std::monostate {};
+struct Known {
+  Name target;
+};
+using IndirectCallInfo = std::variant<Unknown, Trap, Known>;
+
+// Converts indirect calls that target selects between values into ifs over
+// direct calls. For example, consider this input:
+//
+//  (call_ref
+//    (select
+//      (ref.func A)
+//      (ref.func B)
+//      (..condition..)
+//    )
+//  )
+//
+// We'll check if the input falls into such a pattern, and if so, return the new
+// form:
+//
+//  (if
+//    (..condition..)
+//    (call $A)
+//    (call $B)
+//  )
+//
+// If we fail to find the expected pattern, or we decide it is not worth
+// optimizing it for some reason, we return nullptr.
+//
+// If this returns the new form, it will modify the function as necessary,
+// adding new locals etc., which later passes should optimize.
+//
+// |getCallInfo| is given one of the arms of the select and should return an
+// IndirectCallInfo that says what we know about it. We may know nothing, or
+// that it will trap, or that it will go to a known static target.
+template<typename T>
+inline Expression*
+convertToDirectCalls(T* curr,
+                     std::function<IndirectCallInfo(Expression*)> getCallInfo,
+                     Function& func,
+                     Module& wasm) {
+  auto* select = curr->target->template dynCast<Select>();
+  if (!select) {
+    return nullptr;
+  }
+
+  if (select->condition->type == Type::unreachable) {
+    // Leave this for DCE.
+    return nullptr;
+  }
+
+  // Check if we can find useful info for both arms: either known call targets,
+  // or traps.
+  // TODO: support more than 2 targets (with nested selects)
+  auto ifTrueCallInfo = getCallInfo(select->ifTrue);
+  auto ifFalseCallInfo = getCallInfo(select->ifFalse);
+  if (std::get_if<Unknown>(&ifTrueCallInfo) ||
+      std::get_if<Unknown>(&ifFalseCallInfo)) {
+    // We know nothing about at least one arm, so give up.
+    // TODO: Perhaps emitting a direct call for one arm is enough even if the
+    //       other remains indirect?
+    return nullptr;
+  }
+
+  auto& operands = curr->operands;
+
+  // We must use the operands twice, and also must move the condition to
+  // execute first; use locals for them all. While doing so, if we see
+  // any are unreachable, stop trying to optimize and leave this for DCE.
```

Comment:
The comment makes it sound like this is all happening in one loop, but it happens in separate loops. `operandLocals` is not used until down in the next loop.

Reply:
Good catch! This was an issue in the code before it was moved in this PR, but worth fixing. Done.